massa-ledger/src/cursor.rs

```diff
@@ -0,0 +1,132 @@
+use massa_hash::{Hash, HashDeserializer};
+use massa_models::{address::AddressDeserializer, Address};
+use massa_serialization::{Deserializer, SerializeError, Serializer};
+use nom::{sequence::tuple, IResult};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum LedgerCursorStep {
+    Start,
+    Balance,
+    Bytecode,
+    Datastore(Option<Hash>),
+    Finish,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct LedgerCursorStepSerializer;
+
+impl LedgerCursorStepSerializer {
+    fn _new() -> Self {
+        Self
+    }
+}
+
+impl Serializer<LedgerCursorStep> for LedgerCursorStepSerializer {
+    fn serialize(&self, value: &LedgerCursorStep) -> Result<Vec<u8>, SerializeError> {
+        match value {
+            LedgerCursorStep::Start => Ok(vec![0]),
+            LedgerCursorStep::Balance => Ok(vec![1]),
+            LedgerCursorStep::Bytecode => Ok(vec![2]),
+            LedgerCursorStep::Datastore(key) => {
+                let mut bytes = vec![3];
+                if let Some(key) = key {
+                    bytes.extend(key.to_bytes());
+                }
+                Ok(bytes)
+            }
+            LedgerCursorStep::Finish => Ok(vec![4]),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct LedgerCursorStepDeserializer {
+    hash_deserializer: HashDeserializer,
+}
+
+impl LedgerCursorStepDeserializer {
+    fn _new() -> Self {
+        Self {
+            hash_deserializer: HashDeserializer::new(),
+        }
+    }
+}
+
+impl Deserializer<LedgerCursorStep> for LedgerCursorStepDeserializer {
+    fn deserialize<'a>(&self, bytes: &'a [u8]) -> IResult<&'a [u8], LedgerCursorStep> {
+        match bytes[0] {
+            0 => Ok((&bytes[1..], LedgerCursorStep::Start)),
+            1 => Ok((&bytes[1..], LedgerCursorStep::Balance)),
+            2 => Ok((&bytes[1..], LedgerCursorStep::Bytecode)),
+            3 => {
+                if bytes[1..].is_empty() {
+                    Ok((&bytes[1..], LedgerCursorStep::Datastore(None)))
+                } else {
+                    let (rest, key) = self.hash_deserializer.deserialize(&bytes[1..])?;
+                    Ok((rest, LedgerCursorStep::Datastore(Some(key))))
+                }
+            }
+            4 => Ok((&bytes[1..], LedgerCursorStep::Finish)),
+            _ => Err(nom::Err::Error(nom::error::Error::new(
+                bytes,
+                nom::error::ErrorKind::Digit,
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+/// A cursor to iterate through the ledger with different granularity.
+pub struct LedgerCursor(pub Address, pub LedgerCursorStep);
```

Comment:
Can we make these two names fields please?

Reply:
Done